internal/feeder/tiles/tiles_feeder.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tiles is an implementation of a witness feeder for the Tessera tiles log.
+package tiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/transparency-dev/formats/log"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"github.com/transparency-dev/witness/internal/config"
+	"github.com/transparency-dev/witness/internal/feeder"
+	"golang.org/x/mod/sumdb/tlog"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// tileHeight is the tlog tile height.
+	// From: https://developers.google.com/android/binary_transparency/tile
+	tileHeight = 8
+)
+
+// FeedLog periodically feeds checkpoints from the log to the witness.
+// This function returns once the provided context is done.
+func FeedLog(ctx context.Context, l config.Log, w feeder.Witness, c *http.Client, interval time.Duration) error {
+	lURL, err := url.Parse(l.URL)
+	if err != nil {
+		return fmt.Errorf("invalid LogURL %q: %v", l.URL, err)
+	}
+
+	fetchCP := func(ctx context.Context) ([]byte, error) {
+		cpTxt, err := fetch(ctx, c, lURL, "checkpoint")
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch checkpoint: %v", err)
+		}
+		return cpTxt, err
+	}
+	fetchProof := func(ctx context.Context, from, to log.Checkpoint) ([][]byte, error) {
+		if from.Size == 0 {
+			return [][]byte{}, nil
+		}
+		var h [32]byte
+		copy(h[:], to.Hash)
+		tree := tlog.Tree{
+			N:    int64(to.Size),
+			Hash: h,
+		}
+		tr := tileReader{fetch: func(p string) ([]byte, error) {
+			return fetch(ctx, c, lURL, p)
+		}}
+
+		proof, err := tlog.ProveTree(int64(to.Size), int64(from.Size), tlog.TileHashReader(tree, tr))
+		if err != nil {
+			return nil, fmt.Errorf("ProveTree: %v", err)
+		}
+		r := make([][]byte, 0, len(proof))
+		for _, h := range proof {
+			h := h
+			r = append(r, h[:])
+		}
+		klog.V(1).Infof("Fetched proof from %d -> %d: %x", from.Size, to.Size, r)
+		return r, nil
+	}
```

Comment:
I think you might be able to ditch a bunch of stuff in here if you use the `github.com/transparency-dev/trillian-tessera/client` package to interact with the log.

e.g. `client.ProofBuilder` will be able to build the consistency proof for you.

Reply:
I'm not so fluent with Go (yet) to do this, but `serverless_feeder.go` is somewhat similar and I took inspiration from that.

Reply:
Yup, that was an entirely reasonable place to start :)